# Problem 3

Given $f(x) = e^x - 1 - x -\frac{x^2}{2}$,

x = 0 is a zero via the following:

\begin{gather*}

e^0 - 1 - 0 - \frac{0^2}{2} \\

\implies 1 - 1 - 0 - 0 = 0

\end{gather*}


For multiplicity, we need the derivatives.

\begin{gather*}

f(x) = e^x - 1 - x -\frac{x^2}{2} \\
f'(x) = e^x - 1 - x \\
f''(x) = e^x - 2 \\
\newline
f(0) = 0 \\
f'(0) = 0 \\
f''(x) = 1 - 2 = -1

\end{gather*}

This implies that the root multiplicity of the function is 2.



In [13]:
import numpy as np

# modified Newton's method (root multiplicity proof method)
# ASSUMES FUNC IS IN SYMPY FORMAT
def multiplicity_newton(func, p0, f_deriv, s_deriv, n = 100, tol = 0):
    # bc p0 and p1 were found
    i = 1
    q0 = func(p0)
    f0 = f_deriv(p0)
    s0 = s_deriv(p0)
    p1 = p0 - ((q0 * f0) / (f0**2 - q0 * s0))
    # while diff > tol and i < n
    while(abs(p1 - p0) > tol and i < n):
        p0 = p1
        q0 = func(p0)
        f0 = f_deriv(p0)
        s0 = s_deriv(p0)
        p1 = p0 - ((q0 * f0) / (f0**2 - q0 * s0))
        i += 1
    if(i >= n):
        print("Program ran max iterations: ", i)
    else:
        print("Program ran ", i, " iterations")

    # return p2
    return p1


# newton method
# return: p1
# input: function, p0, max iteration, tol
def newton(func, p0, first_deriv, n = 100, tol = 0):
    # bc p0 and p1 were found
    i = 1
    q0 = func(p0)
    f0 = first_deriv(p0)
    p1 = p0 - (q0 / f0)
    # while diff > tol and i < n
    while(abs(p1 - p0) > tol and i < n):
        p0 = p1
        q0 = func(p0)
        p1 = p0 - (q0 / f0)
        i += 1
    if(i >= n):
        print("Program ran max iterations: ", i)
    else:
        print("Program ran ", i, " iterations")

    # return p1
    return p1

# usual newton f(x)
func1 = lambda x: np.exp(x) - 1 -x - (x**2 / 2)
func1_deriv1 = lambda x: np.exp(x) - 1 - x
func1_deriv2 = lambda x: np.exp(x) - 2

# set initial guess
x0 = 1

# modified newton
print("Modified Newton: ", multiplicity_newton(func1, x0, func1_deriv1, func1_deriv2, n = 200000, tol = 10**(-10)))

# usual newton
print("Newton: ", newton(func1, x0, func1_deriv1, n = 200000, tol = 10**(-10)))

Program ran  141412  iterations
Modified Newton:  1.414209775660993e-05
Program ran max iterations:  200000
Newton:  0.0032796387605558263


Using a 200,000 iteration max with a $10^{-10}$ tolerance, the modified Newton method converged with 141,412 iterations while the usual Newton method failed to converge within the max iteration amount. 

This is due to the fact that the given function has a root multiplicity of 2. This implies that there would be roundoff errors while computing the next term while using the usual Newton method.

Using the modified method requires a second derivative, but it does prevent the extra roundoff errors.